packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1777,38 +1790,596 @@ async function prerenderToStream(
+          // During a prospective render the only dynamic thing that can happen is a synchronous dynamic
+          // API access. We expect to have a tracked expression to use for our dynamic error but we fall back
+          // to a generic error if we don't.
+          const dynamicReason = getFirstDynamicReason(dynamicTracking)
+          if (dynamicReason) {
+            throw new DynamicServerError(
+              `Route ${staticGenerationStore.route} couldn't be rendered statically because it used \`${dynamicReason}\`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error`
+            )
+          } else {
+            console.error(
+              'Expected Next.js to keep track of reason for opting out of static rendering but one was not found. This is a bug in Next.js'
+            )
+            throw new DynamicServerError(
+              `Route ${staticGenerationStore.route} couldn't be rendered statically because it used a dynamic API. See more info here: https://nextjs.org/docs/messages/dynamic-server-error`
+            )
+          }
+        } else {
+          // The render didn't explicitly use any Dynamic APIs but it might have IO so we need to retry
+          // the render. We abort the current render here to avoid doing unecessary work.
+          // Keep in mind that while the render is aborted, inflight async ServerComponents can still continue
+          // and might call dynamic APIs.
+          flightController.abort(abortReason)
+        }
+
+        // Reset the prerenderState because we are going to retry the render
+        flightController = new AbortController()
+        dynamicTracking = createDynamicTrackingState(
+          renderOpts.isDebugDynamicAccesses
+        )
+        reactServerIsDynamic = false
+
+        const finalRenderPrerenderStore: PrerenderStore = {
+          // During the final prerender we don't need to track cache access so we omit the signal
+          cacheSignal: null,
+          controller: flightController,
+          dynamicTracking,
+        }
+
+        const finalAttemptRSCPayload = await getRSCPayload(
+          tree,
+          ctx,
+          res.statusCode === 404
+        )
+
+        const reactServerResult = (reactServerPrerenderResult =
+          await createReactServerPrerenderResult(
+            prerenderAndAbortInSequentialTasks(
+              () =>
+                prerenderAsyncStorage.run(
+                  // The store to scope
+                  finalRenderPrerenderStore,
+                  // The function to run
+                  ComponentMod.prerender,
+                  // ... the arguments for the function to run
+                  finalAttemptRSCPayload,
+                  clientReferenceManifest.clientModules,
+                  {
+                    nonce: ctx.nonce,
+                    onError,
+                    onPostpone: () => {
+                      throw new Error('RSC SHOULD NOT POSTPONE WITH DYNAMIC IO')
```

Comment:
same question about these errors... are they invariants or were they used for debugging? if the former we should clean up the error

Reply:
whoops those snuck in somehow